### PR TITLE
[video] Scan art for video extras

### PR DESCRIPTION
**Pull request by CrystalP**
## Description

Detect art for extras in library update / scan new content.

Uses the same patterns as movies:
- generic art type names that will apply to all extras in the folder (ex. poster.jpg, fanart.png, ...)
- art type prefixed by the extra name + "-" applies only to the specific extra name (ex. extra1-poster.jpg).

No art detected: falls back to a default thumb extracted from the extra, same as before.

Some of the scanned art types will never be displayed as skins don't show every art type for extras since at this time they're visible only in the extra selection for playback and in the extras manager dialog.

## Motivation and context

Fill some missing features for video extras.

## How has this been tested?

Added art of different types to the extras directory, some named after the art type, some prefixed with an extra name.

## What is the effect on users?

Ability to set extras art when they're scanned into the library, instead of having to set it manually with the Extras Manager dialog afterwards.

## Screenshots (if appropriate):

With the following file structure:
image

extra1 receives the art extra1-fanart.jpg and extra1-poster.jpg
extra2 and extra3 receive the art of fanart.jpg and poster.jpg
extras in subdir1 and subdir2 receive art from their respective subfolders (if any), not from the parent folder.

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/filesystem/Directory.h [hunk 1]
```diff
@@ -69,8 +69,8 @@ class CDirectory
                            , CFileItemList &items
                            , const CHints &hints);
 
-  using DirectoryEnumerationCallback = std::function<void(const std::shared_ptr<CFileItem>& item)>;
-  using DirectoryFilter = std::function<bool(const std::shared_ptr<CFileItem>& folder)>;
+  using DirectoryEnumerationCallback = std::function<void(std::shared_ptr<CFileItem>& item)>;
+  using DirectoryFilter = std::function<bool(std::shared_ptr<CFileItem>& folder)>;
```
Comment (neo1973) [inline xbmc/filesystem/Directory.h:73]: Why is it necessary, or why is beneficial, to remove the `const` here?

Comment (CrystalP) [reply xbmc/filesystem/Directory.h:73]: That was necessary in AddVideoExtras() due to GetArtwork() modifying the parameter item. Adds art, adds art to its actors, manipulates the video tag a bit (though that part doesn't to survive outside of GetArtwork)

I'm not sure EnumerateDirectory() should have to make a const guarantee. The callers are free to use const auto& item.

Comment (neo1973) [reply xbmc/filesystem/Directory.h:73]: I agree that callers should be able to modify the `CFileItem`. And that's already possible, this PR compiles fine without touching Directory.cpp/h. What the `const` prevents is to change what the `std::shared_ptr` points to, and I believe callers shouldn't be able to change that.

(In hindsight I should have made the filter `std::function<bool(const std::shared_ptr<const CFileItem>& folder)>` as it's a [predicate](https://en.cppreference.com/w/cpp/named_req/Predicate)) .

Comment (CrystalP) [reply xbmc/filesystem/Directory.h]: You're right. I had compilation problems earlier that I resolved that way but it wasn't necessary. Leaving the const in.

File: xbmc/video/VideoDatabase.cpp [hunk 2]
```diff
@@ -12618,12 +12618,15 @@ void CVideoDatabase::AddVideoAsset(VideoDbContentType itemType,
     m_pDS->query(PrepareSQL("SELECT idFile FROM videoversion WHERE idFile = %i", idFile));
 
     if (m_pDS->num_rows() == 0)
-      m_pDS->exec(PrepareSQL("INSERT INTO videoversion VALUES(%i, %i, '%s', %i, %i)", idFile, dbId,
-                             mediaType.c_str(), videoAssetType, idVideoVersion));
+      m_pDS->exec(
+          PrepareSQL("INSERT INTO videoversion (idFile, idMedia, media_type, itemType, idType)"
```
Comment (neo1973) [inline xbmc/video/VideoDatabase.cpp:12622]: Shouldn't there be a space at the end of this string? Not sure if it's relevant.

**Review by neo1973 — APPROVED** [review]
Haven't runtime tested this but code looks good :+1:

**Comment by CrystalP** [comment]
* Addressed the comments. Yes that space was important, and I'm doing more runtime testing before pressing the green button.
* Modified the first commit to follow existing patterns (SetDetailsForXXX family) so the change fits better and artwork addition is included in a db transaction that will improve performance, especially with Sqlite.
* Added a commit for cosmetic changes
* Added a commit for error handling in AddVideoAsset() as its sub functions cascade in multiple places that are not involved with extras art and would make the first commit confusing.
GetArtForItem() was using the return value to indicate whether art was found, which confuses error situations with the absence of artwork > modified along with its callers. The music database had the same design, I didn't change.

**Comment by malvinas2** [comment]
> No art detected: falls back to a default thumb extracted from the extra, same as before.

As a fallback solution, I suggest using the artwork of the default movie, which is usually located in the parent directory of the Extra-directory.

**Comment by CrystalP** [comment]
wiki page modified: https://kodi.wiki/view/Video_extras#Artwork

**Pull request merged**